src/main/java/com/jfrog/ide/common/scan/GraphScanLogic.java

```diff
@@ -0,0 +1,215 @@
+ * The logic uses the graph scan REST API of Xray,
+ * which take into consideration the policy configured in Xray according to given context.
+ * @author tala
+ */
+@SuppressWarnings({"WeakerAccess", "unused"})
+@Getter
+@Setter
+public class GraphScanLogic implements ScanLogic {
+    public static final String MINIMAL_XRAY_VERSION_SUPPORTED_FOR_GRAPH_SCAN = "3.29.0";
+    private ScanCache scanCache;
+    private Log log;
+    private DependencyTree scanResults;
+
+    public GraphScanLogic(ScanCache scanCache, Log log) {
+        this.scanCache = scanCache;
+        this.log = log;
+    }
+
+    /**
+     * Scan and cache components.
+     *
+     * @param server    - JFrog platform server configuration.
+     * @param indicator - Progress bar.
+     * @param quickScan - Quick or full scan.
+     * @return true if the scan completed successfully, false otherwise.
+     */
+    @Override
+    public boolean scanAndCacheArtifacts(ServerConfig server, ProgressIndicator indicator, boolean quickScan, ComponentPrefix prefix, Runnable checkCanceled) throws IOException {
+        // Xray's graph scan API does not support progress indication currently.
+        indicator.setIndeterminate(true);
+        scanResults.setPrefix(prefix.toString());
+        DependencyTree nodesToScan = quickScan ? reduceComponents(scanResults) : scanResults;
+        if (nodesToScan.getChildren().isEmpty()) {
+            log.debug("No components found to scan. '");
+            // No components found to scan
+            return false;
+        }
+        try {
+            // Create Xray client and check version
+            Xray xrayClient = createXrayClientBuilder(server, log).build();
+            if (!isSupportedInXrayVersion(xrayClient)) {
+                return false;
+            }
+            // Start scan
+            log.debug("Starting to scan, sending a dependency graph to Xray");
+            checkCanceled.run();
+            scanComponents(xrayClient, nodesToScan, server.getProject());
+
+            indicator.setFraction(1);
+            log.debug("Saving scan cache...");
+            scanCache.write();
+            log.debug("Scan cache saved successfully.");
+        } catch (CancellationException e) {
+            log.info("Xray scan was canceled.");
+            return false;
+        }
+        return true;
+    }
+
+    /**
+     * Removes components from the dependencies tree, if they already exist in the cache.
+     * This is done to avoid having Xray match the components, for which already have the results.
+     * Please notice that this method checks only the direct dependencies as they are the only ones to
+     * be changed by a developer.
+     *
+     * @param root - The DependencyTree root node.
+     * @return a graph of non cached component for Xray scan.
+     */
+    public DependencyTree reduceComponents(DependencyTree root) {
```

Comment:
Are you sure this logic will work on a multi-module project? If I'm not wrong, the direct children in this case are the projects.